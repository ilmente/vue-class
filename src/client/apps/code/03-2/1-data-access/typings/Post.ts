export interface Post {
    id: string;
    name: string;
    username: string;
    content: string;
    imageSrc: string;
    dislikes: number;
}
