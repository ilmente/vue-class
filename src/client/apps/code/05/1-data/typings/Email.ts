export enum EmailStatus {
    UNREAD = 'unread',
    READ = 'read',
    SPAM = 'spam',
}

export interface EmailInfo {
    id: string;
    sender: string;
    subject: string;
    timestamp: string;
    status: EmailStatus;
}

export interface Email extends EmailInfo {
    content: string;
}
