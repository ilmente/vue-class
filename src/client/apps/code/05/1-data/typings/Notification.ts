export enum NotificationType {
    Default = '',
    Info = 'info',
    Success = 'success',
    Warning = 'warning',
    Danger = 'danger',
}

export interface Notification {
    content: string;
    type: NotificationType;
}
