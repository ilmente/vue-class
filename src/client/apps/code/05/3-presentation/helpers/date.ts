import { formatRelative } from 'date-fns';

export const getRelativeDate = (date?: string): string => date 
    ? formatRelative(new Date(date), new Date())
    : '---';
