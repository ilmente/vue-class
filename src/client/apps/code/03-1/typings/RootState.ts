import { InitialState } from '@typings/InitialState';
import { Post } from './Post';

export interface RootState extends InitialState {
    posts: Post[];
}
