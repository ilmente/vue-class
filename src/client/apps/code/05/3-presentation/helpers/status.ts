import startCase from 'lodash/startCase';
import { EmailStatus } from '../../1-data/typings/Email';

const mofifierMap = {
    [EmailStatus.READ]: 'is-info is-light',
    [EmailStatus.UNREAD]: 'is-success is-light',
    [EmailStatus.SPAM]: 'is-danger is-light',
}

export const getStatusLabel = (status: EmailStatus): string => startCase(status);

export const getStatusModifier = (status: EmailStatus): string => mofifierMap[status];
