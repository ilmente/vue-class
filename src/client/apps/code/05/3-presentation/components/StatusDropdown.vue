<template>
    <div class="dropdown is-right" :class="isOpen && 'is-active'">
        <div class="dropdown-trigger">
            <button
                data-qa-status-dropdown-button
                class="button"
                :class="getStatusModifier(status)"
                @click="onDropdownClick"
                :disabled="isUpdating">
                <span v-if="isUpdating">Updating...</span>
                <span v-else data-qa-status-dropdown-value>{{getStatusLabel(status)}}</span>
            </button>
        </div>
        <div class="dropdown-menu" id="dropdown-menu" role="menu">
            <div class="dropdown-content">
                <a
                    data-qa-status-dropdown-option
                    class="dropdown-item"
                    v-for="(value, index) in statusList"
                    :key="`status-${index}`"
                    :class="isActive(value) && 'is-active'"
                    @click="onDropdownItemClick(value)">
                    {{getStatusLabel(value)}}
                </a>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
    import { Vue, Component, Prop, ModelSync, Watch, Emit } from 'vue-property-decorator';
    import { EmailStatus } from '../../1-data/typings/Email';
    import { getStatusLabel, getStatusModifier } from '../helpers/status';

    @Component({
        name: 'StatusDropdown',
        methods: {
            getStatusLabel,
            getStatusModifier,
        }
    })
    export default class extends Vue {
        @ModelSync('value', 'change', { type: String, default: EmailStatus.UNREAD })
        status!: EmailStatus;

        @Prop({ type: Boolean, default: false })
        isUpdating!: boolean;

        isOpen: boolean = false;

        get statusList() {
            return EmailStatus;
        }

        isActive(status: EmailStatus): boolean {
            return this.status === status;
        }

        @Emit('beforeChange')
        onDropdownClick(): void {
            this.isOpen = !this.isOpen;
        }

        onDropdownItemClick(status: EmailStatus): void {
            this.isOpen = false;
            this.status = status;
        }

        @Watch('isUpdating')
        onUpdating(): void {
            this.isOpen = false;
        }
    }
</script>
