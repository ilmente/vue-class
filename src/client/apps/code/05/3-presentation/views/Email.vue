<template>
    <div class="email p-6">
        <div v-if="isLoading" class="email__progress">
            <progress 
                class="progress is-small is-primary" 
                max="100">
            </progress>
        </div>
        <article v-else-if="email">
            <StatusDropdown 
                class="is-pulled-right"
                :isUpdating="isUpdating"
                :value="email.status"  
                @beforeChange="onStatusBeforeChange"
                @change="onStatusChange" />

            <h1 class="title">{{ email.subject }}</h1>
            <h2 class="subtitle">
                <a>{{ email.sender }}</a>, {{ timestamp }}
            </h2>
            <hr>
            <div class="is-size-4 content" v-html="content"></div>
        </article>
        <div v-else class="email__error">
            <span class="email__icon">ðŸ¤¬ 500</span>
        </div>
    </div>
</template>

<style lang="scss">
    @use 'variables';

    .email {
        position: relative;
        height: 100%;
        overflow: auto;

        &__progress,
        &__error {
            position: relative;
            display: flex;
            align-items: center;
            justify-content: center;
            height: 100%;
        }

        &__icon {
            font-size: 10rem;
        }
    }
</style>

<script lang="ts">
    import { Component, Prop, Watch } from 'vue-property-decorator';
    import { Email, EmailStatus } from '../../1-data/typings/Email';;
    import StoreComponent from '../components/StoreComponent';
    import { getRelativeDate } from '../helpers/date';
    import StatusDropdown from '../components/StatusDropdown.vue';

    const DEFAULT_TIMEOUT = 1500; 

    @Component({
        name: 'Email',
        components: {
            StatusDropdown,
        }
    })
    export default class extends StoreComponent {
        @Prop({ type: String, required: true }) emailId!: string;
        isUpdating: boolean = false;
        timeoutId: ReturnType<typeof setTimeout> | null = null;

        get isLoading(): boolean {
            return this.$store.state.messaging.isLoading && !this.isUpdating;
        }

        get email(): Email | null {
            if (!this.$store.getters['email/exists']) {
                return null;
            }
            
            return this.$store.state.email.current;
        }

        get content(): string {
            return this.$store.getters['email/content'];
        }

        get timestamp(): string {
            return getRelativeDate(this.$store.state.email.current?.timestamp);
        }

        async created(): Promise<void> {
            await this.$store.dispatch('email/load', this.emailId);
            await this.autoUpdateStatus();
        }

        onStatusBeforeChange(): void {
            this.clearAutoUpdateStatus();
        }

        async onStatusChange(status: EmailStatus): Promise<void> {
            this.clearAutoUpdateStatus();
            
            await this.updateStatus(status);
        }

        @Watch('$route')
        async onRouteChange(): Promise<void> {
            this.clearAutoUpdateStatus();

            await this.$store.dispatch('email/load', this.emailId);
            await this.autoUpdateStatus();
        }

        clearAutoUpdateStatus(): void {
            if (!this.timeoutId) {
                return;
            }

            clearTimeout(this.timeoutId);
        }

        async autoUpdateStatus(): Promise<void> {
            if (this.email?.status !== EmailStatus.UNREAD) {
                return;
            }

            this.timeoutId = setTimeout(async () => {
                await this.updateStatus(EmailStatus.READ)
            }, DEFAULT_TIMEOUT);
        }

        async updateStatus(status: EmailStatus): Promise<void> {
            this.isUpdating = true;

            await this.$store.dispatch('email/updateEmailStatus', {
                id: this.email?.id, 
                status,
            });

            this.isUpdating = false;
        }
    }
</script>
