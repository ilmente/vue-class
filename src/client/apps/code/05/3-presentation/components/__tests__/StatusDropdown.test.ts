import StatusDropdown from '../StatusDropdown.vue';
import { mount } from '@vue/test-utils';
import { EmailStatus } from '../../../1-data/typings/Email';
import { getStatusLabel } from '../../helpers/status';

const propsData = {
    value: EmailStatus.READ,
    isUpdating: false,
}

describe('StatusDropdown.vue', () => {
    const wrapper = mount(StatusDropdown, {
        propsData,
    });

    test('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    test('has correct value', () => {
        const option = wrapper.find('.dropdown-item.is-active');
        expect(option.text()).toBe(getStatusLabel(propsData.value));
    });

    test('open the dropdown on click', async () => {
        const button = wrapper.find('button');
        expect(wrapper.classes()).not.toContain('is-active');
        await button.trigger('click');
        expect(wrapper.classes()).toContain('is-active');
    });

    test('emit beforeChange on click', async () => {
        const button = wrapper.find('button');
        await button.trigger('click');
        expect(wrapper.emitted().beforeChange).toBeTruthy();
    });
});
