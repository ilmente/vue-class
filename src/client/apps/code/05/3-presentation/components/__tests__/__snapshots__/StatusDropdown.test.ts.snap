// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusDropdown.vue snapshot 1`] = `
<div
  class="dropdown is-right"
>
  <div
    class="dropdown-trigger"
  >
    <button
      class="button is-info is-light"
    >
      <span>
        Read
      </span>
    </button>
  </div>
   
  <div
    class="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      class="dropdown-content"
    >
      <a
        class="dropdown-item"
      >
        
                Unread
            
      </a>
      <a
        class="dropdown-item is-active"
      >
        
                Read
            
      </a>
      <a
        class="dropdown-item"
      >
        
                Spam
            
      </a>
    </div>
  </div>
</div>
`;
